File: ios/Classes/Internal/FlutterParseQuery.m
//
//  FlutterParseQuery.m
//  flutter_parse
//
//  Created by Alann Maulana on 29/12/18.
//

#import "FlutterParseQuery.h"
#import <Parse/Parse.h>
#import "FlutterParseUtils.h"
#import "FlutterParseObject.h"
#import "PFQueryPrivate.h"
#import "PFRESTQueryCommand.h"

@interface FlutterParseQuery()

@property (strong, nonatomic) PFQuery *query;
@property (assign) BOOL count;

@end

@implementation FlutterParseQuery

- (instancetype)initWithQuery:(PFQuery*) query withCountQuery:(BOOL) count {
    if (self = [super init]) {
        _query = query;
        _count = count;
    }
    
    return self;
}

+ (instancetype)createQuery:(NSDictionary*) dict {
    NSString *className = dict[@"className"];
    if (!className) {
        return nil;
    }
    
    PFQuery *query = [PFQuery queryWithClassName:className];
    PFDecoder *decoder = [PFDecoder objectDecoder];
    
    NSDictionary *where = dict[@"where"];
    [where enumerateKeysAndObjectsUsingBlock:^(id  _Nonnull key, id  _Nonnull obj, BOOL * _Nonnull stop) {
        if ([obj isKindOfClass:[NSDictionary class]]) {
            NSDictionary *clause = obj;
            NSString *type = clause[@"__type"];
            if (type) {
                [query whereKey:key equalTo:[decoder decodeObject:obj]];
            } else {
                [clause enumerateKeysAndObjectsUsingBlock:^(id  _Nonnull keyClause, id  _Nonnull objClause, BOOL * _Nonnull stop) {
                    [self parseQuery:query withKey:key keyClause:keyClause andValue:[decoder decodeObject:objClause]];
                }];
            }
        } else {
            [query whereKey:key equalTo:obj];
        }
    }];
    
    BOOL count = NO;
    NSNumber *numberCount = dict[@"count"];
    if (numberCount) {
        count = numberCount.intValue == 1;
    }
    
    NSNumber *limit = dict[@"limit"];
    if (!limit && limit != 0) {
        query.limit = limit.integerValue;
    }
    
    NSNumber *skip = dict[@"skip"];
    if (!skip) {
        query.skip = skip.integerValue;
    }
    
    NSString *include = dict[@"include"];
    if (include.length > 0) {
        NSArray *split = [include componentsSeparatedByString:@","];
        if (split.count > 0) {
            [query includeKeys:split];
        } else {
            [query includeKey:include];
        }
    }
    
    NSString *order = dict[@"order"];
    if (order.length > 0) {
        NSArray *split = [order componentsSeparatedByString:@","];
        for (NSString *o in split) {
            BOOL desc = [o containsString:@"-"];
            NSString *key = [o stringByReplacingOccurrencesOfString:@"-" withString:@""];
            
            if (desc) {
                [query addDescendingOrder:key];
            } else {
                [query addAscendingOrder:key];
            }
        }
    }
    
    NSString* fields = dict[@"fields"];
    if (fields.length > 0) {
        NSArray *split = [fields componentsSeparatedByString:@","];
        [query selectKeys:split];
    }
    
    return [[FlutterParseQuery alloc] initWithQuery:query withCountQuery:count];
}

+ (void)parseQuery:(PFQuery*) query withKey:(NSString*)key keyClause:(NSString*)keyClause andValue:(id)value {
    PFDecoder *decoder = [PFDecoder objectDecoder];
    
    if ([keyClause isEqualToString:@"whereLessThan"]) {
        [query whereKey:key lessThan:value];
    } else if ([keyClause isEqualToString:@"whereNotEqualTo"]) {
        [query whereKey:key notEqualTo:value];
    } else if ([keyClause isEqualToString:@"whereGreaterThan"]) {
        [query whereKey:key greaterThan:value];
    } else if ([keyClause isEqualToString:@"whereLessThanOrEqualTo"]) {
        [query whereKey:key lessThanOrEqualTo:value];
    } else if ([keyClause isEqualToString:@"whereGreaterThanOrEqualTo"]) {
        [query whereKey:key greaterThanOrEqualTo:value];
    } else if ([keyClause isEqualToString:@"whereContainedIn"] && [value isKindOfClass:[NSArray class]]) {
        [query whereKey:key containedIn:value];
    } else if ([keyClause isEqualToString:@"whereContainsAll"] && [value isKindOfClass:[NSArray class]]) {
        [query whereKey:key containsAllObjectsInArray:value];
    } else if ([keyClause isEqualToString:@"whereFullText"] && [value isKindOfClass:[NSString class]]) {
        // not implemented
    } else if ([keyClause isEqualToString:@"whereNotContainedIn"] && [value isKindOfClass:[NSArray class]]) {
        [query whereKey:key notContainedIn:value];
    } else if ([keyClause isEqualToString:@"whereMatches"] && [value isKindOfClass:[NSDictionary class]]) {
        NSDictionary *dict = value;
        NSString *regex = dict[@"regex"];
        NSString *modifiers = dict[@"modifiers"];
        if (modifiers.length != 0) {
            [query whereKey:key matchesRegex:regex modifiers:modifiers];
        } else {
            [query whereKey:key matchesRegex:regex];
        }
    } else if ([keyClause isEqualToString:@"whereContains"] && [value isKindOfClass:[NSString class]]) {
        [query whereKey:key containsString:value];
    } else if ([keyClause isEqualToString:@"whereStartsWith"] && [value isKindOfClass:[NSString class]]) {
        [query whereKey:key hasPrefix:value];
    } else if ([keyClause isEqualToString:@"whereEndsWith"] && [value isKindOfClass:[NSString class]]) {
        [query whereKey:key hasSuffix:value];
    } else if ([keyClause isEqualToString:@"whereExists"]) {
        [query whereKeyExists:key];
    } else if ([keyClause isEqualToString:@"whereExists"]) {
        [query whereKeyDoesNotExist:key];
    } else if ([keyClause isEqualToString:@"whereDoesNotMatchKeyInQuery"] && [value isKindOfClass:[NSDictionary class]]) {
        NSDictionary *param = value;
        NSString *keyInQuery = param[@"keyInQuery"];
        NSDictionary *queryNotMatches = param[@"query"];
        FlutterParseQuery *parseQuery = [FlutterParseQuery createQuery:queryNotMatches];
        if (parseQuery) {
            [query whereKey:key doesNotMatchKey:keyInQuery inQuery:parseQuery.query];
        }
    } else if ([keyClause isEqualToString:@"whereMatchesKeyInQuery"] && [value isKindOfClass:[NSDictionary class]]) {
        NSDictionary *param = value;
        NSString *keyInQuery = param[@"keyInQuery"];
        NSDictionary *queryMatches = param[@"query"];
        FlutterParseQuery *parseQuery = [FlutterParseQuery createQuery:queryMatches];
        if (parseQuery) {
            [query whereKey:key matchesKey:keyInQuery inQuery:parseQuery.query];
        }
    } else if ([keyClause isEqualToString:@"whereDoesNotMatchQuery"] && [value isKindOfClass:[NSDictionary class]]) {
        NSDictionary *param = value;
        FlutterParseQuery *parseQuery = [FlutterParseQuery createQuery:param];
        if (parseQuery) {
            [query whereKey:key doesNotMatchQuery:parseQuery.query];
        }
    } else if ([keyClause isEqualToString:@"whereMatchesQuery"] && [value isKindOfClass:[NSDictionary class]]) {
        NSDictionary *param = value;
        FlutterParseQuery *parseQuery = [FlutterParseQuery createQuery:param];
        if (parseQuery) {
            [query whereKey:key matchesQuery:parseQuery.query];
        }
    } else if ([keyClause isEqualToString:@"whereNear"] && [value isKindOfClass:[NSDictionary class]]) {
        PFGeoPoint *point = [decoder decodeObject:value];
        [query whereKey:key nearGeoPoint:point];
    } else if ([keyClause isEqualToString:@"whereWithinRadians"] && [value isKindOfClass:[NSDictionary class]]) {
        NSDictionary *param = value;
        PFGeoPoint *point = [decoder decodeObject:param[@"point"]];
        NSNumber *maxDistance = param[@"maxDistance"];
        [query whereKey:key nearGeoPoint:point withinRadians:maxDistance.doubleValue];
    } else if ([keyClause isEqualToString:@"whereWithinGeoBox"] && [value isKindOfClass:[NSArray class]]) {
        NSArray *array = value;
        PFGeoPoint *northWest = [decoder decodeObject:array[0]];
        PFGeoPoint *southEast = [decoder decodeObject:array[1]];
        [query whereKey:key withinGeoBoxFromSouthwest:northWest toNortheast:southEast];
    } else if ([keyClause isEqualToString:@"whereWithinPolygon"] && [value isKindOfClass:[NSArray class]]) {
        NSArray *array = [decoder decodeObject:value];
        [query whereKey:key withinPolygon:array];
    } else if ([keyClause isEqualToString:@"wherePolygonContains"] && [value isKindOfClass:[NSDictionary class]]) {
        PFGeoPoint *point = [decoder decodeObject:value];
        [query whereKey:key polygonContains:point];
    }
}

+ (void)executeAsync:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary *dict = [FlutterParseUtils parsingParams:call.arguments];
    if (!dict) {
        result([FlutterError errorWithCode:[NSString stringWithFormat:@"%ld", kPFErrorInvalidJSON] message:@"invalid parse object" details:nil]);
        return;
    }
    
    FlutterParseQuery *parseQuery = [FlutterParseQuery createQuery:dict];
    if (!parseQuery) {
        result([FlutterError errorWithCode:[NSString stringWithFormat:@"%ld", kPFErrorInvalidJSON] message:@"invalid class name" details:nil]);
        return;
    }
    
//    NSError *e = nil;
//    NSLog(@"QUERY2=%@", [PFRESTQueryCommand findCommandParametersForQueryState:parseQuery.query.state error:&e]);
    
    if (parseQuery.count) {
        [parseQuery.query countObjectsInBackgroundWithBlock:^(int number, NSError * _Nullable error) {
            if (error) {
                result([FlutterError errorWithCode:[NSString stringWithFormat:@"%ld", error.code] message:error.localizedDescription details:nil]);
                return;
            }
            
            result(@(number));
        }];
    } else {
        [parseQuery.query findObjectsInBackgroundWithBlock:^(NSArray * _Nullable objects, NSError * _Nullable error) {
            if (error) {
                result([FlutterError errorWithCode:[NSString stringWithFormat:@"%ld", error.code] message:error.localizedDescription details:nil]);
                return;
            }
            
            NSMutableArray *array = [NSMutableArray array];
            [objects enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
                FlutterParseObject *object = [[FlutterParseObject alloc] initWithParseObject:obj];
                [array addObject:[object dictionary]];
            }];
            result([FlutterParseUtils stringFrom:array]);
        }];
    }
}

@end

File: android/src/main/java/com/parse/FlutterParseQuery.java
package com.parse;

import android.util.Log;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.Arrays;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class FlutterParseQuery {
  private static final String TAG = FlutterParseQuery.class.getSimpleName();
  public static final String QUERY_IN_BACKGROUND = "queryInBackground";

  private final ParseQuery<ParseObject> query;
  private boolean count;

  private FlutterParseQuery(ParseQuery<ParseObject> query, boolean count) {
    this.query = query;
    this.count = count;
  }

  private static FlutterParseQuery createQuery(JSONObject json) {
    String className = json.optString(FlutterParseObject.KEY_CLASS_NAME);
    if (className == null) {
      return null;
    }

    final ParseDecoder decoder = ParseDecoder.get();
    ParseQuery<ParseObject> query = ParseQuery.getQuery(className);

    JSONObject where = json.optJSONObject("where");
    Iterator<String> keys = where.keys();
    while (keys.hasNext()) {
      String key = keys.next();
      Object object = where.opt(key);

      if (object instanceof JSONObject) {
        JSONObject clause = (JSONObject) object;
        String type = clause.optString("__type", null);
        if (type != null) {
          query.whereEqualTo(key, decoder.decode(object));
        } else {
          Iterator<String> keysClause = clause.keys();
          while (keysClause.hasNext()) {
            String keyClause = keysClause.next();
            Object objectClause = clause.opt(keyClause);

            parseQuery(query, key, keyClause, decoder.decode(objectClause));
          }
        }
      } else {
        query.whereEqualTo(key, decoder.decode(object));
      }
    }

    boolean count = json.optInt("count") == 1;

    int limit = json.optInt("limit");
    if (limit != 0) {
      query.setLimit(limit);
    }

    int skip = json.optInt("skip");
    if (skip != 0) {
      query.setSkip(skip);
    }

    String include = json.optString("include");
    if (include.length() > 0) {
      String[] split = include.split(",");
      if (split.length > 0) {
        for (String key : split) {
          query.include(key);
        }
      } else {
        query.include(include);
      }
    }

    String order = json.optString("order");
    if (order.length() > 0) {
      String[] split = order.split(",");
      if (split.length > 0) {
        for (String o : split) {
          boolean desc = o.contains("-");
          String key = o.replace("-", "");
          if (desc) {
            query.addDescendingOrder(key);
          } else {
            query.addAscendingOrder(key);
          }
        }
      } else {
        boolean desc = order.contains("-");
        String key = order.replace("-", "");
        if (desc) {
          query.orderByDescending(key);
        } else {
          query.orderByAscending(key);
        }
      }
    }

    String fields = json.optString("fields");
    if (fields.length() > 0) {
      List<String> list = Arrays.asList(fields.split(","));
      query.selectKeys(list);
    }

    return new FlutterParseQuery(query, count);
  }

  private static void parseQuery(ParseQuery query, String key, String keyClause, Object object) {
    final ParseDecoder decoder = ParseDecoder.get();

    if (keyClause.equals("whereLessThan")) {
      query.whereLessThan(key, object);
    } else if (keyClause.equals("whereNotEqualTo")) {
      query.whereNotEqualTo(key, object);
    } else if (keyClause.equals("whereGreaterThan")) {
      query.whereGreaterThan(key, object);
    } else if (keyClause.equals("whereLessThanOrEqualTo")) {
      query.whereLessThanOrEqualTo(key, object);
    } else if (keyClause.equals("whereGreaterThanOrEqualTo")) {
      query.whereGreaterThanOrEqualTo(key, object);
    } else if (keyClause.equals("whereContainedIn") && object instanceof Collection) {
      query.whereContainedIn(key, (Collection<?>) object);
    } else if (keyClause.equals("whereContainsAll") && object instanceof Collection) {
      //noinspection unchecked
      query.whereContainsAll(key, (Collection<?>) object);
    } else if (keyClause.equals("whereFullText") && object instanceof String) {
      query.whereFullText(key, (String) object);
    } else if (keyClause.equals("whereNotContainedIn") && object instanceof Collection) {
      query.whereNotContainedIn(key, (Collection<?>) object);
    } else if (keyClause.equals("whereMatches") && object instanceof JSONObject) {
      JSONObject param = (JSONObject) object;
      String regex = param.optString("regex");
      String modifiers = param.optString("modifiers");
      if (modifiers.length() != 0) {
        query.whereMatches(key, regex, modifiers);
      } else {
        query.whereMatches(key, regex);
      }
    } else if (keyClause.equals("whereContains") && object instanceof String) {
      query.whereContains(key, (String) object);
    } else if (keyClause.equals("whereStartsWith") && object instanceof String) {
      query.whereStartsWith(key, (String) object);
    } else if (keyClause.equals("whereEndsWith") && object instanceof String) {
      query.whereEndsWith(key, (String) object);
    } else if (keyClause.equals("whereExists")) {
      query.whereExists(key);
    } else if (keyClause.equals("whereDoesNotExist")) {
      query.whereDoesNotExist(key);
    } else if (keyClause.equals("whereDoesNotMatchKeyInQuery") && object instanceof JSONObject) {
      JSONObject param = (JSONObject) object;
      String keyInQuery = param.optString("keyInQuery");
      JSONObject queryNotMatches = param.optJSONObject("query");
      FlutterParseQuery parseQuery = FlutterParseQuery.createQuery(queryNotMatches);
      if (parseQuery != null) {
        //noinspection unchecked
        query.whereDoesNotMatchKeyInQuery(key, keyInQuery, parseQuery.query);
      }
    } else if (keyClause.equals("whereMatchesKeyInQuery") && object instanceof JSONObject) {
      JSONObject param = (JSONObject) object;
      String keyInQuery = param.optString("keyInQuery");
      JSONObject queryMatches = param.optJSONObject("query");
      FlutterParseQuery parseQuery = FlutterParseQuery.createQuery(queryMatches);
      if (parseQuery != null) {
        //noinspection unchecked
        query.whereMatchesKeyInQuery(key, keyInQuery, parseQuery.query);
      }
    } else if (keyClause.equals("whereDoesNotMatchQuery") && object instanceof JSONObject) {
      JSONObject jsonObject = (JSONObject) object;
      FlutterParseQuery parseQuery = FlutterParseQuery.createQuery(jsonObject);
      if (parseQuery != null) {
        //noinspection unchecked
        query.whereDoesNotMatchQuery(key, parseQuery.query);
      }
    } else if (keyClause.equals("whereMatchesQuery") && object instanceof JSONObject) {
      JSONObject jsonObject = (JSONObject) object;
      FlutterParseQuery parseQuery = FlutterParseQuery.createQuery(jsonObject);
      if (parseQuery != null) {
        //noinspection unchecked
        query.whereMatchesQuery(key, parseQuery.query);
      }
    } else if (keyClause.equals("whereNear") && object instanceof JSONObject) {
      ParseGeoPoint point = (ParseGeoPoint) decoder.decode(object);
      query.whereNear(key, point);
    } else if (keyClause.equals("whereWithinRadians") && object instanceof JSONObject) {
      JSONObject jsonObject = (JSONObject) object;
      ParseGeoPoint point = (ParseGeoPoint) decoder.decode(jsonObject.optJSONObject("point"));
      double maxDistance = jsonObject.optDouble("maxDistance");
      query.whereWithinRadians(key, point, maxDistance);
    } else if (keyClause.equals("whereWithinGeoBox") && object instanceof JSONArray) {
      JSONArray array = (JSONArray) object;
      ParseGeoPoint northWest = (ParseGeoPoint) decoder.decode(array.optJSONObject(0));
      ParseGeoPoint southEast = (ParseGeoPoint) decoder.decode(array.optJSONObject(1));
      query.whereWithinGeoBox(key, northWest, southEast);
    } else if (keyClause.equals("whereWithinPolygon") && object instanceof JSONArray) {
      JSONArray array = (JSONArray) object;
      //noinspection unchecked
      List<ParseGeoPoint> points = (List<ParseGeoPoint>) decoder.decode(array);
      query.whereWithinPolygon(key, points);
    } else if (keyClause.equals("wherePolygonContains") && object instanceof JSONObject) {
      JSONObject jsonObject = (JSONObject) object;
      ParseGeoPoint point = (ParseGeoPoint) decoder.decode(jsonObject);
      query.wherePolygonContains(key, point);
    }
  }

  public static void executeAsync(final MethodCall call, final MethodChannel.Result result) {
    JSONObject objectMap = FlutterParseUtils.parsingParams(call.arguments);

    if (objectMap == null) {
      result.error(String.valueOf(ParseException.INVALID_JSON), "invalid parse query", null);
      return;
    }

    final FlutterParseQuery parseQuery = FlutterParseQuery.createQuery(objectMap);
    if (parseQuery == null) {
      result.error(String.valueOf(ParseException.INVALID_JSON), "no className found", null);
      return;
    }

    if (!parseQuery.count) {
      parseQuery.query.findInBackground(new FindCallback<ParseObject>() {
        @Override
        public void done(List<ParseObject> objects, ParseException e) {
          if (e != null) {
            result.error(String.valueOf(e.getCode()), e.getMessage(), null);
            return;
          }

          JSONArray array = new JSONArray();
          for (ParseObject parseObject : objects) {
            FlutterParseObject object = new FlutterParseObject(parseObject);
            array.put(object.toJsonObject());
          }
          result.success(array.toString());
        }
      });
    } else {
      parseQuery.query.countInBackground(new CountCallback() {
        @Override
        public void done(int count, ParseException e) {
          if (e != null) {
            result.error(String.valueOf(e.getCode()), e.getMessage(), null);
            return;
          }

          result.success(count);
        }
      });
    }
  }
}
